### ISSUE TITLE
Vimtex can't execute SumatraPDF viewer

### ISSUE THREAD

**Issue by R1oda** [lervag/vimtex#2459]
### Description

I've  configured SumatraPDF as my vimtex viewer and forward search. However, when I put the command `:VimtexView`, the program isn't enable of execute the PDF viewer.

![Captura de pantalla 2022-08-14 122331](https://user-images.githubusercontent.com/111226848/184534509-83e79ebe-bfc9-430d-9875-cf8304759588.png)

Now I have a 64bits SumatraPDF version installed through chocolatey. Before I installed a 64bits and a 32bits version from its official website, but they haven't worked either.

Moreover, I can't execute the PDF viewer in the cmd of Windows 10 (the OS I'm using). I've tried editing the path variable manually too, but the problem persists.

Finally, now the `:VimtexInfo` command don't work either (more info about this problem is found below, in VimtexInfo section):
![Captura de pantalla 2022-08-14 135532](https://user-images.githubusercontent.com/111226848/184535887-18fd35ae-5f40-47c6-9ca3-a4423c703808.png)

### Steps to reproduce

The minimal init.vim:
![Captura de pantalla 2022-08-14 122416](https://user-images.githubusercontent.com/111226848/184534543-40dcebdf-7102-4b88-ab55-3dbf0ccc476b.png)

Some useful nvim-qt  info:
![Captura de pantalla 2022-08-14 122218](https://user-images.githubusercontent.com/111226848/184534582-9cf5194a-d9b6-419a-bb03-6954a84d6d55.png)

### Expected behavior

Once compiled the tex file into a PDF (until here vimtex works fine), I write the `:VimtexView` command and it opens another windows with only SumatraPDF showing the PDF.

### Actual behavior

I write the `:VimtexView` command and I get the error message that I described above. 

### Do you use a latexmkrc file?

No

### VimtexInfo

Before I wrote this issue, `:VimtexInfo` worked fine, but now I find this error and I don't understand which file is talking about. The error message is above, in the first section.

If it's necessary, I can open another issue about this problem.

**Comment by lervag** [comment]
> I've configured SumatraPDF as my vimtex viewer and forward search. However, when I put the command `:VimtexView`, the program isn't enable of execute the PDF viewer.
> 
> Moreover, I can't execute the PDF viewer in the cmd of Windows 10 (the OS I'm using). I've tried editing the path variable manually too, but the problem persists.

Let's start there. I mean, if we can't run SumatraPDF from a cmd prompt, then it _will_ be hard to do it in neovim.

Did you read the docs, in particular `:help vimtex-faq-windows` and `:help vimtex-view-sumatrapdf`?

I believe you did not _properly_ edit the PATH variable. You _need_ to do so. When it works, you should be able to run SumatraPDF.exe from a cmd prompt.

> Finally, now the `:VimtexInfo` command don't work either (more info about this problem is found below, in VimtexInfo section): ...

This is because you are using a non-UTF8 system. I don't really know how to "fix" that on the VimTeX side, so I instead encourage you to change your system somehow to use UTF8. I have _no_ idea how to do that either, sorry. You can do `:VimtexClearCache ALL` to clear all caches - it should allow you do run `:VimtexInfo` again, at least once. And you can use `let g:vimtex_cache_persistent = 0` to disable caching, which will avoid the problem at the cost of reduced performance.

> The minimal init.vim:

You don't need `let g:tex_flavor` with VimTeX.

**Comment by R1oda** [comment]
First, the command :VimtexClearCache ALL didn't work at all: once I pressed enter, the command line stayed the same and :VimtexInfo continued not running. Fortunately, adding  let g:vimtex_cache_persistent = 0 to the minimal init.vim solved the problem, so here is my VimtexInfo:

![Captura de pantalla 2022-08-18 121218](https://user-images.githubusercontent.com/111226848/185387950-33813ca5-87f8-49a2-88d2-c76bed8d48b2.png)

However, I'm a bit concerned about the reduction of perfomance. Can I undo this setting when I don't need :VimtexInfo?

Moreover, I've read more carefully the :help vimtex-faq-windows section. From this reading I can conclude that I shouldn't have any problem with Perl since I'd already installed it. Thus, the error is caused by not editing properly the path variable or by some PowerShell strange behaviour I haven't  corrected yet.

In the first case, I can't see why my path variable isn't properly edit. The help section also offers a link from where I can install a software to editing the path variable, but after doing so it didn't seem I edited the path better than doing from the windows control panel. In any case, I upload an image showing the path of the SumatraPDF executable (highlighted in blue at the right) and the new paths I added to the path variable (highlighted in the middle window):

![Captura de pantalla 2022-08-18 133238](https://user-images.githubusercontent.com/111226848/185391942-be01e30d-bbcc-4c16-8279-b16ee6929bec.png)

In the second case, I don't understand what is the default 'shell' settings. Does it refers to the windows cmd?

Thanks for the replies.

**Comment by lervag** [comment]
> First, the command :VimtexClearCache ALL didn't work at all: once I pressed enter, the command line stayed the same and :VimtexInfo continued not running.

After the command, please restart Vim and VimTeX.

> Fortunately, adding let g:vimtex_cache_persistent = 0 to the minimal init.vim solved the problem, so here is my VimtexInfo:

Good, thanks!

> However, I'm a bit concerned about the reduction of perfomance. Can I undo this setting when I don't need :VimtexInfo?

When things work, you should be able to remove `g:vimtex_cache_persistent`. Open a new issue on this specific thing if you find it does not work.

> In any case, I upload an image showing the path of the SumatraPDF executable (highlighted in blue at the right) and the new paths I added to the path variable (highlighted in the middle window):

That's not right. We are not asking you to create a new variable, we are asking you to update the `PATH` variable. It is one of the _existing_ variables, and you need to add the path `C:\Program Files\SumatraPDF` to that variable.

![image](https://user-images.githubusercontent.com/998898/185598436-b04b19bd-3603-4f45-9cb1-c6481aa1d530.png)

> In the second case, I don't understand what is the default 'shell' settings. Does it refers to the windows cmd?

No, it refers to the Vim option `'shell'`, see `:help 'shell'`. You problably have not changed it, so this is not important here.

**Comment by R1oda** [comment]
Finally, it worked! It was the first time I edited the environment variables and I though that the cmd searched for a variable name equal to some string input and then ran the correspondent directory. Now I see it's more complex than that.

Since SumatraPDF already opens, I've could configure the inverse search and works fine. 
The window popups are a bit annoying, but after writing the code that reduces them I suppose there aren't any better ways to avoid the popups.

I've also configured the forward search, but whenever I run the command :VimtexView I get the next message in the pdf viewer:

![P](https://user-images.githubusercontent.com/111226848/185625895-e032a5c7-f27a-4d4c-b632-a73956838897.png)

(it states an error loading the directory where the tex file is stored followed, apparently, of the command that runs the forward search)

**Comment by lervag** [comment]
Perhaps you could show me the configuration you use?

**Comment by R1oda** [comment]
I'm using the same init.vim code that I sent in the first message if that is what do you mean. If not, could you detail a bit more your question?

**Comment by lervag** [comment]
Sorry, I did not mean to be unclear. The thing is, it is somewhat unclear to me exactly how your configuration looks. So, can you use the following configuration (it may be the same as you are already using, but please dobbel check):

```vim
let g:vimtex_view_general_viewer = 'SumatraPDF'
let g:vimtex_view_general_options
      \ = '-reuse-instance -forward-search @tex @line @pdf'
```

Then:

1. Open your LaTeX document.
2. Do `\ll` to compile.
3. Do `\lv` to open viewer (if it did not open automatically).
4. Do `\lv` to do forward search.
5. Do `\li` to show the output from `:VimtexInfo`. Copy that output here (preferably verbatim, instead of screenshot; but if that's hard, then a screenshot also works).

Follow the above steps, and comment on any errors or problems along the way.

**Comment by lervag** [comment]
> I've just followed the above steps and I got two error messages from SumatraPDF. The first one is exactly the same that appears in the last screenshot I sent, and appeared when I compiled the code and the pdf viewer opened automatically. The second one appeared when I tried to execute forward search and it talked about some synchronization problem.

Can you take a screenshot of the second error message as well?

> Here I show the `:VimtexInfo` output: ...

Thanks. Can you open a windows command prompt (or powershell) and run these commands from the folder `C:\Users\ricks\Desktop\Latex`:

```sh
# First, try this:
SumatraPDF -reuse-instace -forward-search "C:\Users\ricks\Desktop\Latex\p.tex" 18 "C:\Users\ricks\Desktop\Latex/p.pdf"

# Next, try this:
SumatraPDF -reuse-instace -forward-search "C:\Users\ricks\Desktop\Latex\p.tex" 18 "C:\Users\ricks\Desktop\Latex\p.pdf"
```

> Finally, I have a question about the continuous mode of the compiler. Should it save my code every time I change it? I ask it because if I change the code before compiling and then I compile, the compiler compiles the older code and not the new one. In other words, do I have to save the code every time I make a change in order to update the pdf file?

Continuous mode means that `latexmk` will stay in the background and run the compiler (e.g. pdflatex or lualatex or similar) whenever you save the document. It does not compile everytime you change any thing, only after you save.

**Comment by R1oda** [comment]
Here is the seccond error message:
![Captura de pantalla 2022-08-25 094232](https://user-images.githubusercontent.com/111226848/186607702-1f59681f-9777-47eb-86ee-690071994e17.png)

It states: "Synchronization file cannot be opened"

I've also tried your commands in the Windows cmd, but neither of them have worked. In fact, I've obtained the exact two error messages I obtained before.

**Comment by lervag** [comment]
I'm really sorry, but I have _no_ idea what is going on here. Hopefully, someone else may be watching who knows more about Windows and is perhaps able to help. One thing: Does it work if you simplify the options specified by `g:vimtex_view_general_options`?

**Comment by R1oda** [comment]
I don't know exactly how to simplify g:vimtex_view_general_options.

Not worry, you have already helped a lot. If I find the solution in another website or by own, I'll write it here. Thanks for your help.

**Comment by lervag** [comment]
With simplify, I mean things like:

```vim
let g:vimtex_view_general_viewer = 'SumatraPDF'

" Simpler alternatives for following option:
let g:vimtex_view_general_options = ''
let g:vimtex_view_general_options = '-forward-search @tex @line @pdf'
let g:vimtex_view_general_options = '-reuse-instance'
```

> Not worry, you have already helped a lot. If I find the solution in another website or by own, I'll write it here. Thanks for your help.

Thanks! And yes, feel free to update here if you figure things out!

**Comment by xarthurx** [comment]
SumatraPDF does not work here either.

In my case, I haven't used `vimtex` for around half a year and started to write something recently.
Before, neovim + vimtex + sumatrapdf works perfectly.

Now, the viewer simply won't start.

- `SumatraPDF` and `Neovim` both 64-bit.
- `SumatraPDF` is in the `PATH`.
- using simplified `let g:vimtex_view_general_options = ''` won't help.

Though for me open `SumatraPDF` manually and continue with my work won't bother me a lot, but hopefully this issue could be resolved at some point.

**Comment by lervag** [comment]
@xarthurx It is not clear to me that your problem is the same as that of @R1oda. Notice that for @R1oda, VimTeX _does_ open SumatraPDF, it is just that the pdf does not properly open and instead, some error message appears.

Unless you are really experiencing the same situation, please create a new issue and explain in more detail how you've setup VimTeX (i.e. your config) and SumatraPDF. Also, ensure that you have read `:help vimtex-faq-windows` as well as `:help vimtex-view-sumatrapdf`.

**Comment by xarthurx** [comment]
> @xarthurx It is not clear to me that your problem is the same as that of @R1oda. Notice that for @R1oda, VimTeX _does_ open SumatraPDF, it is just that the pdf does not properly open and instead, some error message appears.
> 
> Unless you are really experiencing the same situation, please create a new issue and explain in more detail how you've setup VimTeX (i.e. your config) and SumatraPDF. Also, ensure that you have read `:help vimtex-faq-windows` as well as `:help vimtex-view-sumatrapdf`.

Thanks for the reply.
Yeah, it is different. I use the minimal example and realize that it must be some vim settings that prevent the view from starting.
I'll figure out by myself. Thanks.

**Comment by lervag** [comment]
Sounds good; good luck! Feel free to open an issue if you want assistance.

**Comment by NicoStellwag** [comment]
> > @xarthurx It is not clear to me that your problem is the same as that of @R1oda. Notice that for @R1oda, VimTeX _does_ open SumatraPDF, it is just that the pdf does not properly open and instead, some error message appears.
> > Unless you are really experiencing the same situation, please create a new issue and explain in more detail how you've setup VimTeX (i.e. your config) and SumatraPDF. Also, ensure that you have read `:help vimtex-faq-windows` as well as `:help vimtex-view-sumatrapdf`.
> 
> Thanks for the reply. Yeah, it is different. I use the minimal example and realize that it must be some vim settings that prevent the view from starting. I'll figure out by myself. Thanks.

@xarthurx Did you figure this out by any chance?
I'm currently facing exactly the same issue, but it doesn't seem like my (neo)vim installation has any issues with starting other external programs.

**Comment by NicoStellwag** [comment]
> Yeah, my issue is from a plugin I used and I didn't set it correctly. It is better to start with the basic `.vimrc` settings and add your settings little by little.

Alright, thanks!

**Comment by caiosaboia** [comment]
So I had a similar problem and I just put a direct path to the sumatrapdf.exe file and it solved it. I don't know if the issue is still open, but it's worth a try.

in init.lua (or init.vim) try this:
"vim.g.vimtex_view_general_viewer = "C:\\Users\\youruser\\AppData\\Local\\SumatraPDF\\SumatraPDF.exe"
"


(I use Windows 11 and Neovim)

**Comment by lervag** [comment]
> So I had a similar problem and I just put a direct path to the sumatrapdf.exe file and it solved it.

Or you could adjust the PATH environment variable as explained in the docs (see the second bullet in the list at `:help vimtex-faq-windows`). At least, I believe that it should work and that it would make your workaround unnecessary.
